aeneas/src/ir/VariantSolver.v3

```diff
@@ -0,0 +1,171 @@
+// Copyright 2024 Virgil authors. All rights reserved.
+// See LICENSE for details of Apache 2.0 license.
+
+class VariantNorm extends TypeNorm {
+	def fields: Array<VariantField>;
+
+	def tag: VariantField;
+	var tagValue: int = -1;
+
+	// for each field, this is the start + end index when ClassAlloc is called
+	var fieldRanges: Array<(int, int)>;
+
+	new(oldType: Type, newType: Type, sub: Array<Type>, fields, tag)
+		super(oldType, newType, sub) {}
+
+	// represent empty variant as tag
+	def isEnum() -> bool { return size == 1 && tag != null && fields.length == 0; }
+	def isEnumCase() -> bool { return size == 1 && tag != null && tagValue >= 0 && fields.length == 0; }
+
+	def tagType() -> IntType { return IntType.!(tag.tn.first()); }
+	def tagIndex() -> int { return tag.indexes[0]; }
+	
+	// represent single-case, unboxed variant without its tag
+	def isTagless() -> bool { return tag == null; }
+
+	def render(buf: StringBuilder) -> StringBuilder {
+		buf.put3("%q ## %q |%d|", oldType.render, newType.render, size);
+		buf.puts(" [");
+		for (i < sub.length) {
+			if (i > 0) buf.csp();
+			sub[i].render(buf);
+		}
+		buf.puts("] {");
+		if (tag != null) {
+			buf.put1("tag: %q", tag.render);
+		}
+		for (i < fields.length) {
+			if (i > 0 || tag != null) buf.csp();
+			fields[i].render(buf);
+		}
+		buf.puts("}");
+		return buf;
+	}
+}
+
+// Metadata about a variant's fields before/after normalization
+class VariantField {
+	def tn: TypeNorm;
+	def indexes: Array<int>; // Indexes into the variant's representation to extract the field from
+
+	new(tn, indexes) {}
+
+	def render(buf: StringBuilder) -> StringBuilder {
+		buf.puts("(");
+		for (i < indexes.length) {
+			if (i > 0) buf.csp();
+			buf.put1("#%d", indexes[i]);
+		}
+		buf.puts(")");
+		if (tn != null) buf.put1(": [%q]", tn.newType.render);
+		return buf;
+	}
+}
+
+def ON_STACK = -1;
+
+class VariantSolver(nc: NormalizerConfig, rn: ReachabilityNormalizer, verbose: bool) {
+	def normVariant(t: Type, rc: RaClass) -> TypeNorm {
```

Comment:
Can you documents the methods of this class? It's the important interface between the `Normalizer` and the solver.